WIP: Configuration refactoring

jendakol opened the pull request:
At this moment, this PR is about checking I'm going the right way.  
The plan is to remove configuration from Config and make the core module completely independent on it, and add the configuration from Config afterwards by another module. Now I finished the first phase; removed the Ficus and Config and moved default values to case classes. I'd like to have a confirmation, that the current state is OK before I proceed to adding the Pureconfig module.  
Thx.

jakubjanecek commented:
LGTM

hanny24 commented:
I have one one comment. Please try to be consistent with config class names. Either they all should have `Config` suffix, or none of them. My preference would be to include the suffix.

jendakol commented:
@hanny24 I kind of agree, but that will break backward compatibility (it's not the only place and now it's probably good time to do that, all at one moment). Should I do it?

hanny24 commented:
You are breaking it anyway, right?

jendakol commented:
As I said; it's not the only place (which break it) and from some POV it's better to break it now on multiple places than later multiple times...

jendakol commented:
I have implemented all the functionality.  
I want to emphasize:
1. The API provided by `pureconfig` module is compatible with previous versions - kind of :smiley: It's not guaranteed, however unit tests passed without any change. I'll have to do some more testing to know more, but I expect strong majority of people won't recognize any difference (in terms of their *.conf files; there are some minor differences in types that the API return).
1. I made a huge refactoring of configuration internals (e.g. `DefaultRabbitMQClientFactory` lost 200 lines!), no changes were made to _actual code_ nor behavior - this PR is all about configuration.
1. I used Pureconfig as a configuration library instead of Ficus. It didn't bring me any (dis-)advantages, but it's much more widely spread.

This is not done yet:
1. Updated README - I will do it as a last thing before release, to prevent it being out-of-date.
1. Configuration of SSL -  it should support more types types than just JKS and also receive custom SSLContext as an input.

These are things to discuss:
1. Configuration of SSL - I believe this is something we could extract to some library (recently discussed with @jakubjanecek) which will be then used here. I hate copying this code between projects :slightly_smiling_face:

I'll keep this PR as _draft_ until the SSL problem is solved, however I consider rest of the code to be final. For me, a very good sign was that test didn't require any change after the refactoring - that IMHO means it's backward compatible (as much as possible).

jakubjanecek commented on pureconfig/src/main/scala/com/avast/clients/rabbitmq/pureconfig/PureconfigRabbitMQConnection.scala:
```diff
@@ -0,0 +1,136 @@
+import scala.language.higherKinds
+import scala.reflect.ClassTag
+
+trait PureconfigRabbitMQConnection[F[_]] {
```
I don't get this trait. There is nothing PureConfig-related in it except that it uses that library internally. I don't think this should be a trait at all, what you need is an extension method of `RabbitMqConnection` that just loads the configuration using PureConfig and passes the configuration as case class to the underlying "original" implementation. Nothing more. Trait basically represents some kind of abstraction and this is no such thing.

jendakol replied:
Solved in person.

jakubjanecek commented on pureconfig/src/main/scala/com/avast/clients/rabbitmq/pureconfig/Implicits.scala:
```diff
@@ -0,0 +1,96 @@
+
+trait Implicits {
+  // use camelCase names (see https://pureconfig.github.io/docs/overriding-behavior-for-case-classes.html#field-mappings)
+  private implicit def hint[T]: ProductHint[T] = ProductHint[T](ConfigFieldMapping(CamelCase, CamelCase))
```
Shouldn't this be configurable? You are forcing everyone to use your convention.

jendakol replied:
OK, good idea, however the default has to be this, to keep backward compatibility. Not sure how to achieve that, will think about that.

jakubjanecek replied:
Yes

jakubjanecek commented on extras/src/main/scala/com/avast/clients/rabbitmq/extras/testing/testing.scala:
```diff
@@ -99,11 +82,11 @@ package object testing {
   }
 
-  private def fkFToSyncIO[F[_]](implicit F: Effect[F]): FunctionK[F, SyncIO] = new FunctionK[F, SyncIO] {
+  private[rabbitmq] def fkFToSyncIO[F[_]](implicit F: Effect[F]): FunctionK[F, SyncIO] = new FunctionK[F, SyncIO] {
```
Why are these things needed? If the client is really "finally tagless" than you can instantiate it to any effect data type you want so why would you do that and immediately convert it to some other data type?

jakubjanecek replied:
And I see it's supposed to be for testing but even for that I don't understand it.

jendakol replied:
Yes. This is supposed to be suitable for testing. In past, it offered `RabbitMQConnection[Eval]`, but now I had to change it to `SyncIO`. The point was to introduce _blocking_ behavior (e.g. `producer.send` would not  have to be _run_ but would behave like a blocking call), unfortunately since migrating to `Resource` it's not possible (`Resource won't accept anything that's not `Bracket` etc.) So the current state is that you have `SyncIO` and you have to call `unsafeRunSync()` anyway :-/ so it maybe have lost it's sense. From my current POV, I have no problem deleting it, which you won't oppose I guess? :smiley:

jakubjanecek commented on extras-pureconfig/src/main/scala/com/avast/clients/rabbitmq/extras/pureconfig/testing/testing.scala:
```diff
@@ -0,0 +1,75 @@
+
+import scala.language.higherKinds
+
+package object testing {
```
I don't understand this module. It's called `pureconfig-extras` but it actually does conversion of effect data types. It seems strange.

jendakol replied:
I don't like the name either (I forgot to mention it), even though it doesn't do what you're saying :smiley:

jendakol replied:
Better toe be discussed in person.

jendakol replied:
Solved in person

sideeffffect commented on pureconfig/src/main/scala/com/avast/clients/rabbitmq/pureconfig/PureconfigImplicits.scala:
```diff
@@ -0,0 +1,94 @@
+import com.rabbitmq.client.RecoveryDelayHandler
+import org.slf4j.event.Level
+
+private[pureconfig] class PureconfigImplicits(implicit namingConvention: NamingConvention = CamelCase) {
```
> `CamelCase`

why don't we just use kebab case, which is PureConfig default?

jakubjanecek replied:
`CamelCase` is chosen for backwards compatibility because that's what we've been using in the past.

sideeffffect commented on pureconfig/src/main/scala/com/avast/clients/rabbitmq/pureconfig/PureconfigImplicits.scala:
```diff
@@ -0,0 +1,94 @@
+import com.rabbitmq.client.RecoveryDelayHandler
+import org.slf4j.event.Level
+
+private[pureconfig] class PureconfigImplicits(implicit namingConvention: NamingConvention = CamelCase) {
```
also, why do we even need this at all? PureConfig loading should work automatically (at least for vast majority of use cases) without any need for manual/semiautomatic implicits

jakubjanecek replied:
We have decided in the past that fully automatic mode is not going to be used. There are reasons of course. One of them is speed of compilation because the fully automatic mode sometimes means that readers are derived again and again instead of one being derived in the `val`.  Another one is that this allows you to have at least some control over the readers. For example you might want to have a special reader for `String` (e.g. hex-string) so you could put it here and be sure that it will work like this for your library, in fully automatic mode this is not possible and it depends on your imports whether `String` is read as string or hex-string. The few lines of code you need to write seem to be on the safer side even though it's less ergonomic to do.

jendakol commented:
FYI this PR is almost done, but it's waiting for SSL configuration. I've been considering using ssl-config from Lightbend but it doesn't met my expectations (not without some refactoring) so now I have to decide how to solve it.

sideeffffect commented on core/src/main/scala/com/avast/clients/rabbitmq/RabbitMQConnection.scala:
```diff
@@ -117,63 +70,15 @@ object RabbitMQConnection extends StrictLogging {
-        val connectionConfig = config.as[RabbitMQConnectionConfig](FakeConfigRootName)
-        val connection = createConnection(connectionConfig, blockingExecutor, connectionListener, channelListener, consumerListener)
-
+      consumerListener: ConsumerListener = DefaultListeners.DefaultConsumerListener): Resource[F, DefaultRabbitMQConnection[F]] = {
```
this should be
```scala
: Resource[F, RabbitMQConnection[F]] = {
```

jendakol replied:
I agree

sideeffffect commented on core/src/main/scala/com/avast/clients/rabbitmq/DefaultRabbitMQClientFactory.scala:
```diff
@@ -31,262 +23,87 @@ private[rabbitmq] object DefaultRabbitMQClientFactory extends LazyLogging {
         blocker: Blocker,
         monitor: Monitor,
         consumerListener: ConsumerListener,
         readAction: DeliveryReadAction[F, A])(implicit timer: Timer[F], cs: ContextShift[F]): DefaultRabbitMQConsumer[F] = {
```
should return
```scala
: RabbitMQConsumer[F] = {
```

jendakol replied:
The same as below.

sideeffffect commented on core/src/main/scala/com/avast/clients/rabbitmq/DefaultRabbitMQClientFactory.scala:
```diff
@@ -31,262 +23,87 @@ private[rabbitmq] object DefaultRabbitMQClientFactory extends LazyLogging {
-        channelFactoryInfo: RabbitMQConnectionInfo,
+        connectionInfo: RabbitMQConnectionInfo,
         blocker: Blocker,
         monitor: Monitor)(implicit cs: ContextShift[F]): DefaultRabbitMQPullConsumer[F, A] = {
```
should be
```scala
: DefaultRabbitMQPullConsumer[F, A] = {
```

jendakol replied:
I disagree, this is private object which serves as factory for a specific implementation and as such it's allowed to return specific type.

sideeffffect commented on pureconfig/src/main/scala/com/avast/clients/rabbitmq/pureconfig/PureconfigImplicits.scala:
```diff
@@ -0,0 +1,90 @@
+import com.rabbitmq.client.RecoveryDelayHandler
+import org.slf4j.event.Level
+
+private[pureconfig] class PureconfigImplicits(implicit namingConvention: NamingConvention = CamelCase) {
```
this entity is usually called `implicits`, so I suggest to follow this convention

jendakol replied:
Solved in person.

jendakol commented:
I think the work is done. After waiting several weeks for SSL configuration to resolve, today @jakubjanecek and @sideeffffect convinced me to remove it completely and take `Option[SSLContext]` in factory method as dependency (the same way as `ExecutionContext` is taken) so I did it.  It's source-compatible with previous version in terms of it won't fail, however it has to be stressed in documentation, that existing SSL coknfiguration in `Config` won't be taken into account anymore.  
Now it's ready to be merged into master and for release of another RC, heading towards stable release soon (I'd wait for some time however, since there were some changes that could possibly affect the performance in this branch).
